--- lazarus/install/lazarus.desktop.orig	2015-07-28 11:55:15.349067745 +0100
+++ lazarus/install/lazarus.desktop	2015-07-28 11:55:29.626882382 +0100
@@ -4,13 +4,8 @@
 Exec=startlazarus %f
 Terminal=false
 Type=Application
-Icon=lazarus.png
-Categories=Application;IDE;Development;GTK;GUIDesigner;
+Icon=lazarus
+Categories=IDE;Development;GTK;GUIDesigner;
 StartupWMClass=Lazarus
 MimeType=text/x-pascal;text/lazarus-project-source;text/lazarus-project-information;text/lazarus-form;text/lazarus-resource;text/lazarus-package;text/lazarus-package-link;text/lazarus-code-inlay;
 Keywords=editor;Pascal;IDE;FreePascal;fpc;Design;Designer;
-[Property::X-KDE-NativeExtension]
-Type=QString
-Value=.pas
-X-Ubuntu-Gettext-Domain=desktop_kdelibs
-
